brave/src/main/java/brave/Tracer.java

```diff
@@ -369,27 +365,48 @@ public Span nextSpan(TraceContextOrSamplingFlags extracted) {
     return _toSpan(decorateContext(flags, traceIdHigh, traceId, localRootId, spanId, 0L, extra));
   }
 
-  /** Converts the context to a Span object after decorating it for propagation */
+  /**
+   * Converts the context to a Span object after decorating it for propagation.
+   *
+   * <p>This api is not advised for routine use. It is better to hold a reference to a span created
+   * elsewhere vs rely on implicit lookups.
+   */
   public Span toSpan(TraceContext context) {
-    if (context == null) throw new NullPointerException("context == null");
-    if (isDecorated(context)) return _toSpan(context);
-
-    return _toSpan(decorateContext(
+    // Re-use a pending context if present: This ensures reference consistency on Span.context()
+    TraceContext pendingContext = swapForPendingContext(context);
+    if (pendingContext != null) return _toSpan(pendingContext);
+
+    // There are a few known scenarios for the context to be absent from the pending map:
+    // * Created by a separate tracer (localRootId set)
+    // * Resurrected span from a garbage collection (localRootId set)
```

Comment:
this one is very hard to grok, but basically if the input context is the same ids as the one garbage collected, but a different instance, it is effectively resurrection from the POV of the user (as they wouldn't know about ref swapping anyway). I can remove this or revise this comment if confusing

Reply:
sgtm thanks